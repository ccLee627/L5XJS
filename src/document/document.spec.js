const Document = require("./index");
const fs = require("fs");
const path = require("path");

describe("Document Class", () => {
    describe("New Instance", () => {
        it("Accepts Proper Input", () => {
            const fn = arg => () => new Document(arg);

            expect(fn("./__test-data__/fakeFile.L5X")).toThrow();
            expect(fn(57)).toThrow();
            expect(fn("./__test-data__/datatype.L5X")).not.toThrow();
            expect(fn()).not.toThrow();
        });

        it("Initializes with Desired Document Model", () => {
            const doc1 = new Document();
            const doc2 = new Document("./__test-data__/datatype.L5X");

            expect(doc1._dom).toMatchSnapshot();
            expect(doc2._dom).toMatchSnapshot();
        });
    });

    describe("Static Methods", () => {
        test("isDocument: Returns Appropriate Judgement", () => {
            const doc = new Document();
            const fake = {};

            expect(Document.isDocument(doc)).toBeTruthy();
            expect(Document.isDocument(fake)).toBeFalsy();
        });
    });

    describe("Public Methods", () => {
        let docFull;
        let docEmpty;

        beforeEach(() => {
            docFull = new Document("./__test-data__/datatype.L5X");
            docEmpty = new Document();
        });

        test("get dom: Gets Mutable Dom Reference", () => {
            expect(docFull.dom).toMatchSnapshot();
        });

        test("find: Finds Document Elements and Returns New Document", () => {
            expect(docFull.find("nothing")).toBeNull();
            expect(docEmpty.find("nothing")).toBeNull();

            expect(docFull.find("Controller")).not.toBeNull();
            expect(docFull.find("Controller")).toMatchSnapshot();

            expect(docEmpty.find("Controller")).not.toBeNull();
            expect(docEmpty.find("Controller")).toMatchSnapshot();

            expect(docFull.find("Member", { Name: "Push" })).not.toBeNull();
            expect(docEmpty.find("Member", { Name: "Push" })).toBeNull();

            expect(docFull.find("Member", { Name: "Push" })).toMatchSnapshot();
            expect(docFull.find("Member")).toMatchSnapshot();

            expect(
                docFull.find("Controller")[0].find("Member", { Name: "Push" })
            ).toMatchSnapshot();
            expect(docFull.find("Controller")[0].find("Member", { name: "Push" })).toBeNull();
        });

        test("find: Rejects Invalid Inputs", () => {
            const fn = (type, attr, tree) => () => docFull.find(type, attr, tree);

            expect(fn(12, null, null)).toThrow();
            expect(fn("Controller", null, null)).not.toThrow();
            expect(fn("Controller", 12, null)).toThrow();
            expect(fn("Controller", { Name: "another" }, null)).not.toThrow();
            expect(fn("Controller", { Name: "another" }, 12)).toThrow();
            expect(fn("Controller", { Name: "another" }, {})).not.toThrow();
        });

        test("replace: Throws on Invalid Inputs", () => {
            const fn = arg => () => docEmpty.replace(arg);

            expect(fn({ key: "value" })).toThrow();
            expect(fn(docFull)).not.toThrow();
        });

        test("replace: Successfully Updates Document", () => {
            let push = docFull.find("Member", { Name: "Push" })[0];

            push.dom.attributes.Name = "BetterPush";

            expect(docFull.replace(push)).toBeTruthy();
            expect(docFull._dom).toMatchSnapshot();

            push._dom.attributes = null;
            expect(docFull.replace(push)).toBeFalsy();
        });

        test("export: Exported Data is Correct", () => {
            docFull.export("./__test-data__/generated_datatype.L5X");

            let original = fs.readFileSync(
                path.resolve(__dirname, "./__test-data__/datatype.L5X"),
                "utf8"
            );

            let generated = fs.readFileSync(
                path.resolve(__dirname, "./__test-data__/generated_datatype.L5X"),
                "utf8"
            );

            original = original.replace(/\s/g, "");
            generated = generated.replace(/\s/g, "");

            expect(original === generated).toBeTruthy();

            fs.unlinkSync(path.resolve(__dirname, "./__test-data__/generated_datatype.L5X"));
        });
    });
});
