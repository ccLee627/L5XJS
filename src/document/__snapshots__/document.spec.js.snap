// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document Class New Instance Initializes with Desired Document Model 1`] = `
Object {
  "declaration": Object {
    "attributes": Object {
      "encoding": "UTF-8",
      "standalone": "yes",
      "version": "1.0",
    },
  },
  "elements": Array [
    Object {
      "attributes": Object {
        "ContainsContext": true,
        "SchemaRevision": "1.0",
      },
      "elements": Array [
        Object {
          "attributes": Object {
            "Use": "context",
          },
          "elements": Array [],
          "name": "Controller",
          "type": "element",
        },
      ],
      "name": "RSLogix5000Content",
      "type": "element",
    },
  ],
}
`;

exports[`Document Class New Instance Initializes with Desired Document Model 2`] = `
Object {
  "declaration": Object {
    "attributes": Object {
      "encoding": "UTF-8",
      "standalone": "yes",
      "version": "1.0",
    },
  },
  "elements": Array [
    Object {
      "attributes": Object {
        "ContainsContext": "true",
        "ExportDate": "Tue Aug 29 09:06:17 2017",
        "ExportOptions": "References NoRawData L5KData DecoratedData Context Dependencies ForceProtectedEncoding AllProjDocTrans",
        "Owner": "Operator",
        "SchemaRevision": "1.0",
        "SoftwareRevision": "28.00",
        "TargetName": "PB_OneButton",
        "TargetType": "DataType",
      },
      "elements": Array [
        Object {
          "attributes": Object {
            "Name": "TRW_10673_Final",
            "Use": "Context",
          },
          "elements": Array [
            Object {
              "attributes": Object {
                "Use": "Context",
              },
              "elements": Array [
                Object {
                  "attributes": Object {
                    "Class": "User",
                    "Family": "NoFamily",
                    "Name": "PB_OneButton",
                    "Use": "Target",
                  },
                  "elements": Array [
                    Object {
                      "elements": Array [
                        Object {
                          "attributes": Object {
                            "DataType": "SINT",
                            "Dimension": "0",
                            "ExternalAccess": "Read/Write",
                            "Hidden": "true",
                            "Name": "ZZZZZZZZZZPB_OneButt0",
                            "Radix": "Decimal",
                          },
                          "name": "Member",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {
                            "BitNumber": "0",
                            "DataType": "BIT",
                            "Dimension": "0",
                            "ExternalAccess": "Read/Write",
                            "Hidden": "false",
                            "Name": "Push",
                            "Radix": "Decimal",
                            "Target": "ZZZZZZZZZZPB_OneButt0",
                          },
                          "elements": Array [
                            Object {
                              "elements": Array [
                                Object {
                                  "cdata": "*",
                                  "type": "cdata",
                                },
                              ],
                              "name": "Description",
                              "type": "element",
                            },
                          ],
                          "name": "Member",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {
                            "BitNumber": "1",
                            "DataType": "BIT",
                            "Dimension": "0",
                            "ExternalAccess": "Read/Write",
                            "Hidden": "false",
                            "Name": "State",
                            "Radix": "Decimal",
                            "Target": "ZZZZZZZZZZPB_OneButt0",
                          },
                          "elements": Array [
                            Object {
                              "elements": Array [
                                Object {
                                  "cdata": "*",
                                  "type": "cdata",
                                },
                              ],
                              "name": "Description",
                              "type": "element",
                            },
                          ],
                          "name": "Member",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {
                            "BitNumber": "2",
                            "DataType": "BIT",
                            "Dimension": "0",
                            "ExternalAccess": "Read/Write",
                            "Hidden": "false",
                            "Name": "Visibility",
                            "Radix": "Decimal",
                            "Target": "ZZZZZZZZZZPB_OneButt0",
                          },
                          "elements": Array [
                            Object {
                              "elements": Array [
                                Object {
                                  "cdata": "*",
                                  "type": "cdata",
                                },
                              ],
                              "name": "Description",
                              "type": "element",
                            },
                          ],
                          "name": "Member",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {
                            "BitNumber": "3",
                            "DataType": "BIT",
                            "Dimension": "0",
                            "ExternalAccess": "Read/Write",
                            "Hidden": "false",
                            "Name": "Ind",
                            "Radix": "Decimal",
                            "Target": "ZZZZZZZZZZPB_OneButt0",
                          },
                          "elements": Array [
                            Object {
                              "elements": Array [
                                Object {
                                  "cdata": "*",
                                  "type": "cdata",
                                },
                              ],
                              "name": "Description",
                              "type": "element",
                            },
                          ],
                          "name": "Member",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {
                            "BitNumber": "4",
                            "DataType": "BIT",
                            "Dimension": "0",
                            "ExternalAccess": "Read/Write",
                            "Hidden": "false",
                            "Name": "Interlock",
                            "Radix": "Decimal",
                            "Target": "ZZZZZZZZZZPB_OneButt0",
                          },
                          "elements": Array [
                            Object {
                              "elements": Array [
                                Object {
                                  "cdata": "*",
                                  "type": "cdata",
                                },
                              ],
                              "name": "Description",
                              "type": "element",
                            },
                          ],
                          "name": "Member",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {
                            "BitNumber": "5",
                            "DataType": "BIT",
                            "Dimension": "0",
                            "ExternalAccess": "Read/Write",
                            "Hidden": "false",
                            "Name": "ONS",
                            "Radix": "Decimal",
                            "Target": "ZZZZZZZZZZPB_OneButt0",
                          },
                          "elements": Array [
                            Object {
                              "elements": Array [
                                Object {
                                  "cdata": "*",
                                  "type": "cdata",
                                },
                              ],
                              "name": "Description",
                              "type": "element",
                            },
                          ],
                          "name": "Member",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {
                            "DataType": "TIMER",
                            "Dimension": "0",
                            "ExternalAccess": "Read/Write",
                            "Hidden": "false",
                            "Name": "Tmr",
                            "Radix": "NullType",
                          },
                          "elements": Array [
                            Object {
                              "elements": Array [
                                Object {
                                  "cdata": "*",
                                  "type": "cdata",
                                },
                              ],
                              "name": "Description",
                              "type": "element",
                            },
                          ],
                          "name": "Member",
                          "type": "element",
                        },
                      ],
                      "name": "Members",
                      "type": "element",
                    },
                  ],
                  "name": "DataType",
                  "type": "element",
                },
              ],
              "name": "DataTypes",
              "type": "element",
            },
          ],
          "name": "Controller",
          "type": "element",
        },
      ],
      "name": "RSLogix5000Content",
      "type": "element",
    },
  ],
}
`;

exports[`Document Class Public Methods Find: Finds Document Elements and Returns New Document 1`] = `
Array [
  Document {
    "_dom": Object {
      "attributes": Object {
        "Name": "TRW_10673_Final",
        "Use": "Context",
      },
      "elements": Array [
        Object {
          "attributes": Object {
            "Use": "Context",
          },
          "elements": Array [
            Object {
              "attributes": Object {
                "Class": "User",
                "Family": "NoFamily",
                "Name": "PB_OneButton",
                "Use": "Target",
              },
              "elements": Array [
                Object {
                  "elements": Array [
                    Object {
                      "attributes": Object {
                        "DataType": "SINT",
                        "Dimension": "0",
                        "ExternalAccess": "Read/Write",
                        "Hidden": "true",
                        "Name": "ZZZZZZZZZZPB_OneButt0",
                        "Radix": "Decimal",
                      },
                      "name": "Member",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {
                        "BitNumber": "0",
                        "DataType": "BIT",
                        "Dimension": "0",
                        "ExternalAccess": "Read/Write",
                        "Hidden": "false",
                        "Name": "Push",
                        "Radix": "Decimal",
                        "Target": "ZZZZZZZZZZPB_OneButt0",
                      },
                      "elements": Array [
                        Object {
                          "elements": Array [
                            Object {
                              "cdata": "*",
                              "type": "cdata",
                            },
                          ],
                          "name": "Description",
                          "type": "element",
                        },
                      ],
                      "name": "Member",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {
                        "BitNumber": "1",
                        "DataType": "BIT",
                        "Dimension": "0",
                        "ExternalAccess": "Read/Write",
                        "Hidden": "false",
                        "Name": "State",
                        "Radix": "Decimal",
                        "Target": "ZZZZZZZZZZPB_OneButt0",
                      },
                      "elements": Array [
                        Object {
                          "elements": Array [
                            Object {
                              "cdata": "*",
                              "type": "cdata",
                            },
                          ],
                          "name": "Description",
                          "type": "element",
                        },
                      ],
                      "name": "Member",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {
                        "BitNumber": "2",
                        "DataType": "BIT",
                        "Dimension": "0",
                        "ExternalAccess": "Read/Write",
                        "Hidden": "false",
                        "Name": "Visibility",
                        "Radix": "Decimal",
                        "Target": "ZZZZZZZZZZPB_OneButt0",
                      },
                      "elements": Array [
                        Object {
                          "elements": Array [
                            Object {
                              "cdata": "*",
                              "type": "cdata",
                            },
                          ],
                          "name": "Description",
                          "type": "element",
                        },
                      ],
                      "name": "Member",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {
                        "BitNumber": "3",
                        "DataType": "BIT",
                        "Dimension": "0",
                        "ExternalAccess": "Read/Write",
                        "Hidden": "false",
                        "Name": "Ind",
                        "Radix": "Decimal",
                        "Target": "ZZZZZZZZZZPB_OneButt0",
                      },
                      "elements": Array [
                        Object {
                          "elements": Array [
                            Object {
                              "cdata": "*",
                              "type": "cdata",
                            },
                          ],
                          "name": "Description",
                          "type": "element",
                        },
                      ],
                      "name": "Member",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {
                        "BitNumber": "4",
                        "DataType": "BIT",
                        "Dimension": "0",
                        "ExternalAccess": "Read/Write",
                        "Hidden": "false",
                        "Name": "Interlock",
                        "Radix": "Decimal",
                        "Target": "ZZZZZZZZZZPB_OneButt0",
                      },
                      "elements": Array [
                        Object {
                          "elements": Array [
                            Object {
                              "cdata": "*",
                              "type": "cdata",
                            },
                          ],
                          "name": "Description",
                          "type": "element",
                        },
                      ],
                      "name": "Member",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {
                        "BitNumber": "5",
                        "DataType": "BIT",
                        "Dimension": "0",
                        "ExternalAccess": "Read/Write",
                        "Hidden": "false",
                        "Name": "ONS",
                        "Radix": "Decimal",
                        "Target": "ZZZZZZZZZZPB_OneButt0",
                      },
                      "elements": Array [
                        Object {
                          "elements": Array [
                            Object {
                              "cdata": "*",
                              "type": "cdata",
                            },
                          ],
                          "name": "Description",
                          "type": "element",
                        },
                      ],
                      "name": "Member",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {
                        "DataType": "TIMER",
                        "Dimension": "0",
                        "ExternalAccess": "Read/Write",
                        "Hidden": "false",
                        "Name": "Tmr",
                        "Radix": "NullType",
                      },
                      "elements": Array [
                        Object {
                          "elements": Array [
                            Object {
                              "cdata": "*",
                              "type": "cdata",
                            },
                          ],
                          "name": "Description",
                          "type": "element",
                        },
                      ],
                      "name": "Member",
                      "type": "element",
                    },
                  ],
                  "name": "Members",
                  "type": "element",
                },
              ],
              "name": "DataType",
              "type": "element",
            },
          ],
          "name": "DataTypes",
          "type": "element",
        },
      ],
      "name": "Controller",
      "type": "element",
    },
    "class_id": "ba7cc09d99f4a84810458bc89abf612f",
  },
]
`;

exports[`Document Class Public Methods Find: Finds Document Elements and Returns New Document 2`] = `
Array [
  Document {
    "_dom": Object {
      "attributes": Object {
        "Use": "context",
      },
      "elements": Array [],
      "name": "Controller",
      "type": "element",
    },
    "class_id": "ba7cc09d99f4a84810458bc89abf612f",
  },
]
`;

exports[`Document Class Public Methods Find: Finds Document Elements and Returns New Document 3`] = `
Array [
  Document {
    "_dom": Object {
      "attributes": Object {
        "BitNumber": "0",
        "DataType": "BIT",
        "Dimension": "0",
        "ExternalAccess": "Read/Write",
        "Hidden": "false",
        "Name": "Push",
        "Radix": "Decimal",
        "Target": "ZZZZZZZZZZPB_OneButt0",
      },
      "elements": Array [
        Object {
          "elements": Array [
            Object {
              "cdata": "*",
              "type": "cdata",
            },
          ],
          "name": "Description",
          "type": "element",
        },
      ],
      "name": "Member",
      "type": "element",
    },
    "class_id": "ba7cc09d99f4a84810458bc89abf612f",
  },
]
`;

exports[`Document Class Public Methods Find: Finds Document Elements and Returns New Document 4`] = `
Array [
  Document {
    "_dom": Object {
      "attributes": Object {
        "DataType": "SINT",
        "Dimension": "0",
        "ExternalAccess": "Read/Write",
        "Hidden": "true",
        "Name": "ZZZZZZZZZZPB_OneButt0",
        "Radix": "Decimal",
      },
      "name": "Member",
      "type": "element",
    },
    "class_id": "ba7cc09d99f4a84810458bc89abf612f",
  },
  Document {
    "_dom": Object {
      "attributes": Object {
        "BitNumber": "0",
        "DataType": "BIT",
        "Dimension": "0",
        "ExternalAccess": "Read/Write",
        "Hidden": "false",
        "Name": "Push",
        "Radix": "Decimal",
        "Target": "ZZZZZZZZZZPB_OneButt0",
      },
      "elements": Array [
        Object {
          "elements": Array [
            Object {
              "cdata": "*",
              "type": "cdata",
            },
          ],
          "name": "Description",
          "type": "element",
        },
      ],
      "name": "Member",
      "type": "element",
    },
    "class_id": "ba7cc09d99f4a84810458bc89abf612f",
  },
  Document {
    "_dom": Object {
      "attributes": Object {
        "BitNumber": "1",
        "DataType": "BIT",
        "Dimension": "0",
        "ExternalAccess": "Read/Write",
        "Hidden": "false",
        "Name": "State",
        "Radix": "Decimal",
        "Target": "ZZZZZZZZZZPB_OneButt0",
      },
      "elements": Array [
        Object {
          "elements": Array [
            Object {
              "cdata": "*",
              "type": "cdata",
            },
          ],
          "name": "Description",
          "type": "element",
        },
      ],
      "name": "Member",
      "type": "element",
    },
    "class_id": "ba7cc09d99f4a84810458bc89abf612f",
  },
  Document {
    "_dom": Object {
      "attributes": Object {
        "BitNumber": "2",
        "DataType": "BIT",
        "Dimension": "0",
        "ExternalAccess": "Read/Write",
        "Hidden": "false",
        "Name": "Visibility",
        "Radix": "Decimal",
        "Target": "ZZZZZZZZZZPB_OneButt0",
      },
      "elements": Array [
        Object {
          "elements": Array [
            Object {
              "cdata": "*",
              "type": "cdata",
            },
          ],
          "name": "Description",
          "type": "element",
        },
      ],
      "name": "Member",
      "type": "element",
    },
    "class_id": "ba7cc09d99f4a84810458bc89abf612f",
  },
  Document {
    "_dom": Object {
      "attributes": Object {
        "BitNumber": "3",
        "DataType": "BIT",
        "Dimension": "0",
        "ExternalAccess": "Read/Write",
        "Hidden": "false",
        "Name": "Ind",
        "Radix": "Decimal",
        "Target": "ZZZZZZZZZZPB_OneButt0",
      },
      "elements": Array [
        Object {
          "elements": Array [
            Object {
              "cdata": "*",
              "type": "cdata",
            },
          ],
          "name": "Description",
          "type": "element",
        },
      ],
      "name": "Member",
      "type": "element",
    },
    "class_id": "ba7cc09d99f4a84810458bc89abf612f",
  },
  Document {
    "_dom": Object {
      "attributes": Object {
        "BitNumber": "4",
        "DataType": "BIT",
        "Dimension": "0",
        "ExternalAccess": "Read/Write",
        "Hidden": "false",
        "Name": "Interlock",
        "Radix": "Decimal",
        "Target": "ZZZZZZZZZZPB_OneButt0",
      },
      "elements": Array [
        Object {
          "elements": Array [
            Object {
              "cdata": "*",
              "type": "cdata",
            },
          ],
          "name": "Description",
          "type": "element",
        },
      ],
      "name": "Member",
      "type": "element",
    },
    "class_id": "ba7cc09d99f4a84810458bc89abf612f",
  },
  Document {
    "_dom": Object {
      "attributes": Object {
        "BitNumber": "5",
        "DataType": "BIT",
        "Dimension": "0",
        "ExternalAccess": "Read/Write",
        "Hidden": "false",
        "Name": "ONS",
        "Radix": "Decimal",
        "Target": "ZZZZZZZZZZPB_OneButt0",
      },
      "elements": Array [
        Object {
          "elements": Array [
            Object {
              "cdata": "*",
              "type": "cdata",
            },
          ],
          "name": "Description",
          "type": "element",
        },
      ],
      "name": "Member",
      "type": "element",
    },
    "class_id": "ba7cc09d99f4a84810458bc89abf612f",
  },
  Document {
    "_dom": Object {
      "attributes": Object {
        "DataType": "TIMER",
        "Dimension": "0",
        "ExternalAccess": "Read/Write",
        "Hidden": "false",
        "Name": "Tmr",
        "Radix": "NullType",
      },
      "elements": Array [
        Object {
          "elements": Array [
            Object {
              "cdata": "*",
              "type": "cdata",
            },
          ],
          "name": "Description",
          "type": "element",
        },
      ],
      "name": "Member",
      "type": "element",
    },
    "class_id": "ba7cc09d99f4a84810458bc89abf612f",
  },
]
`;

exports[`Document Class Public Methods Find: Finds Document Elements and Returns New Document 5`] = `
Array [
  Document {
    "_dom": Object {
      "attributes": Object {
        "BitNumber": "0",
        "DataType": "BIT",
        "Dimension": "0",
        "ExternalAccess": "Read/Write",
        "Hidden": "false",
        "Name": "Push",
        "Radix": "Decimal",
        "Target": "ZZZZZZZZZZPB_OneButt0",
      },
      "elements": Array [
        Object {
          "elements": Array [
            Object {
              "cdata": "*",
              "type": "cdata",
            },
          ],
          "name": "Description",
          "type": "element",
        },
      ],
      "name": "Member",
      "type": "element",
    },
    "class_id": "ba7cc09d99f4a84810458bc89abf612f",
  },
]
`;

exports[`Document Class Public Methods toString: Returns Accurate String 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"yes\\"?><RSLogix5000Content SchemaRevision=\\"1.0\\" SoftwareRevision=\\"28.00\\" TargetName=\\"PB_OneButton\\" TargetType=\\"DataType\\" ContainsContext=\\"true\\" Owner=\\"Operator\\" ExportDate=\\"Tue Aug 29 09:06:17 2017\\" ExportOptions=\\"References NoRawData L5KData DecoratedData Context Dependencies ForceProtectedEncoding AllProjDocTrans\\"><Controller Use=\\"Context\\" Name=\\"TRW_10673_Final\\"><DataTypes Use=\\"Context\\"><DataType Use=\\"Target\\" Name=\\"PB_OneButton\\" Family=\\"NoFamily\\" Class=\\"User\\"><Members><Member Name=\\"ZZZZZZZZZZPB_OneButt0\\" DataType=\\"SINT\\" Dimension=\\"0\\" Radix=\\"Decimal\\" Hidden=\\"true\\" ExternalAccess=\\"Read/Write\\"/><Member Name=\\"Push\\" DataType=\\"BIT\\" Dimension=\\"0\\" Radix=\\"Decimal\\" Hidden=\\"false\\" Target=\\"ZZZZZZZZZZPB_OneButt0\\" BitNumber=\\"0\\" ExternalAccess=\\"Read/Write\\"><Description><![CDATA[*]]></Description></Member><Member Name=\\"State\\" DataType=\\"BIT\\" Dimension=\\"0\\" Radix=\\"Decimal\\" Hidden=\\"false\\" Target=\\"ZZZZZZZZZZPB_OneButt0\\" BitNumber=\\"1\\" ExternalAccess=\\"Read/Write\\"><Description><![CDATA[*]]></Description></Member><Member Name=\\"Visibility\\" DataType=\\"BIT\\" Dimension=\\"0\\" Radix=\\"Decimal\\" Hidden=\\"false\\" Target=\\"ZZZZZZZZZZPB_OneButt0\\" BitNumber=\\"2\\" ExternalAccess=\\"Read/Write\\"><Description><![CDATA[*]]></Description></Member><Member Name=\\"Ind\\" DataType=\\"BIT\\" Dimension=\\"0\\" Radix=\\"Decimal\\" Hidden=\\"false\\" Target=\\"ZZZZZZZZZZPB_OneButt0\\" BitNumber=\\"3\\" ExternalAccess=\\"Read/Write\\"><Description><![CDATA[*]]></Description></Member><Member Name=\\"Interlock\\" DataType=\\"BIT\\" Dimension=\\"0\\" Radix=\\"Decimal\\" Hidden=\\"false\\" Target=\\"ZZZZZZZZZZPB_OneButt0\\" BitNumber=\\"4\\" ExternalAccess=\\"Read/Write\\"><Description><![CDATA[*]]></Description></Member><Member Name=\\"ONS\\" DataType=\\"BIT\\" Dimension=\\"0\\" Radix=\\"Decimal\\" Hidden=\\"false\\" Target=\\"ZZZZZZZZZZPB_OneButt0\\" BitNumber=\\"5\\" ExternalAccess=\\"Read/Write\\"><Description><![CDATA[*]]></Description></Member><Member Name=\\"Tmr\\" DataType=\\"TIMER\\" Dimension=\\"0\\" Radix=\\"NullType\\" Hidden=\\"false\\" ExternalAccess=\\"Read/Write\\"><Description><![CDATA[*]]></Description></Member></Members></DataType></DataTypes></Controller></RSLogix5000Content>"`;

exports[`Document Class Public Methods toString: Returns Accurate String 2`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"yes\\"?><RSLogix5000Content SchemaRevision=\\"1.0\\" ContainsContext=\\"true\\"><Controller Use=\\"context\\"/></RSLogix5000Content>"`;
